### PR TITLE
Optuna hparams search

### LINKED ISSUE #103: Hyperparameter Search for AM Datasets + Approaches
Up to this point, we have integrated 6 different AM datasets and two model approaches. However, the model performance does not look so well when using the default config parameters. So we need to find better hyperparameters.

Parameter candidates:
 - learning rate
 - transformer base model

What else? Let us extend this list. Maybe we can get some inspiration from: https://github.com/google-research/tuning_playbook

We should have individual issues for all the combinations and link them in the table below.

| datatasets    | pipeline<br/>naive: token classification > RE text classification | joined<br/>generative (BART), see #26 |
|---------------|:----------------------------------------------------------------:|:----------------------------:|
| aae2          |                                                                  |                              |
| abstrct       |                                                                  |                              |
| argmicro      |                                                                  |                              |
| cdcp          |                                                                  |                              |
| sciarg        |                                                                  |                              |
| scidtb_argmin |                                                                  |                              |

### PR DESCRIPTION
**Pull request by jk-tripathy**
Can be merged once https://github.com/ArneBinder/pie-document-level/pull/108 is completed.

Hparam Seach can be run with: `python src/train.py model=token_classification_with_seq2seq_encoder_and_crf experiment=sciarg_adus_sam_paper hparams_search=full_sam_paper`

### CODE REVIEW

File: configs/hparams_search/full_sam_paper.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      seed: 1234
+      n_startup_trials: 5 # number of random sampling runs before optimization starts
+
+    # define range of hyperparameters
+    # More information here : https://hydra.cc/docs/plugins/optuna_sweeper/#search-space-configuration
+    params:
+      base_model: choice(allenai/scibert_scivocab_uncased, allenai/longformer-base-4096, allenai/longformer-scico)
+      model.learning_rate: choice(5e-5, 3e-5, 2e-5, 3e-4, 1e-4)
+      +model.task_learning_rate: choice(5e-5, 3e-5, 2e-5, 3e-4, 1e-4)
+      model.freeze_base_model: choice(true, false)
```
Comment (ArneBinder) [inline configs/hparams_search/full_sam_paper.yaml]: I think it makes sense to have a different hparams_search config for `freeze_base_model=true` and one for `false`. Freezing the base model will *require* additional parameters (e.g. in the seq2seqencoder) because the single linear classifier layer (+crf) should not be able to capture the relevant semantics for the task. This may not be the case when fine-tuning the whole base transformer model.

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Initial feedback.

File: configs/hparams_search/full_sam_paper.yaml (hunk 1 above)
Comment (ArneBinder) [reply configs/hparams_search/full_sam_paper.yaml]: Also: the experiment config `sciarg_adus_sam_paper` is especially build (and we did already some hyperparameter search last year) for the case of `freeze_base_model=true`. But we tried just with `allenai/scibert_scivocab_uncased`. So, some slightly different hps (hyperparameters) may work better, but especially for the other models.

Comment (jk-tripathy) [reply configs/hparams_search/full_sam_paper.yaml]: Right that makes sense, ill make two separate files for the search

Comment (jk-tripathy) [reply configs/hparams_search/full_sam_paper.yaml]: > Also: the experiment config `sciarg_adus_sam_paper` is especially build (and we did already some hyperparameter search last year) for the case of `freeze_base_model=true`. But we tried just with `allenai/scibert_scivocab_uncased`. So, some slightly different hps (hyperparameters) may work better, but especially for the other models.

Right got it

File: configs/hparams_search/frozen_full.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      seed: 1234
+      n_startup_trials: 5 # number of random sampling runs before optimization starts
+
+    # define range of hyperparameters
+    # More information here : https://hydra.cc/docs/plugins/optuna_sweeper/#search-space-configuration
+    params:
+      taskmodule.tokenizer_name_or_path: choice(allenai/scibert_scivocab_uncased, allenai/longformer-base-4096, allenai/longformer-scico)
+      model.learning_rate: choice(5e-5, 3e-5, 2e-5, 3e-4, 1e-4)
+      +model.task_learning_rate: choice(5e-5, 3e-5, 2e-5, 3e-4, 1e-4)
+      model.freeze_base_model: true
```
Comment (ArneBinder) [inline configs/hparams_search/frozen_full.yaml]: I guess single values (no range or choice) need to be defined just outside the `hydra` section (e.g. at the very bottom of the file).

File: configs/hparams_search/unfozen_trunc.yaml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    # docs: https://optuna.readthedocs.io/en/stable/reference/samplers.html
+    sampler:
+      _target_: optuna.samplers.TPESampler
+      seed: 1234
+      n_startup_trials: 5 # number of random sampling runs before optimization starts
+
+    # define range of hyperparameters
+    # More information here : https://hydra.cc/docs/plugins/optuna_sweeper/#search-space-configuration
+    params:
+      taskmodule.tokenizer_name_or_path:w: choice(allenai/scibert_scivocab_uncased, allenai/longformer-base-4096, allenai/longformer-scico)
```
Comment (ArneBinder) [inline configs/hparams_search/unfozen_trunc.yaml]: remove the "w"

**Review by ArneBinder — CHANGES_REQUESTED** [review]
I guess the values look fine (but I just glanced over them). But please add small descriptions to the beginning of the three new `hprams_search` config files describing what they are about  (and fix the spelling in `unfozen_trunc.yaml`). Also add a small description to this PR saying whats added so far (note that the other configs will be added via the other PR and do not need to be described here).

EDIT: Please try with the biggest model what batch size works before starting the full hparams search. Would be unfortunate if it breaks in the middle.

FYI I rebased the branch and also the respective branch in `pie-models`. Please update and `pip uninstall pie-models` & re-install from `requirements.txt`.

File: configs/hparams_search/frozen_full.yaml (hunk 2 above)
Comment (jk-tripathy) [reply configs/hparams_search/frozen_full.yaml]: got it

File: configs/hparams_search/unfozen_trunc.yaml (hunk 3 above)
Comment (jk-tripathy) [reply configs/hparams_search/unfozen_trunc.yaml]: ah my bad, ill change that

File: configs/hparams_search/frozen_full.yaml (hunk 2 above)
Comment (jk-tripathy) [reply configs/hparams_search/frozen_full.yaml]: i checked this btw, does not seem to matter where the param is defined

File: configs/hparams_search/bert_frozen_with_seq2seq_enc.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# @package _global_
+
+# example hyperparameter optimization of some experiment with Optuna:
+# python train.py -m hparams_search=conll2003_optuna experiment=conll2003
```
Comment (ArneBinder) [inline configs/hparams_search/bert_frozen_with_seq2seq_enc.yaml]: Replace with short description that gives the motivation (why do we need this?) and content (e.g. different bert variants + s2s encoder, because...) of this config and adjust the example command (also in the other configs)

File: configs/model/seq2seq_encoder/lstm.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+type: lstm
```
Comment (ArneBinder) [inline configs/model/seq2seq_encoder/lstm.yaml]: add a short description, e.g.
```suggestion
# two layer Bi-LSTM with dropout
type: lstm
```

File: configs/model/seq2seq_encoder/many_encoders.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+type: sequential
+lin0:
+  type: linear
+  out_features: 2048
+act0:
+  type: tanh
```
Comment (ArneBinder) [inline configs/model/seq2seq_encoder/many_encoders.yaml]: add a short description (see lstm example), maybe say that `lin0` and `act0` are arbitrary identifiers, but its order matters (defines how the sequential is constructed)

File: configs/model/seq2seq_encoder/none.yaml [hunk 7]
```diff
@@ -0,0 +1 @@
+type: none
```
Comment (ArneBinder) [inline configs/model/seq2seq_encoder/none.yaml]: add short description (e.g. empty dummy encoder to use no sequence-to-sequence encoder at all)

**Review by ArneBinder — CHANGES_REQUESTED** [review]
Some minor things.

**Review by ArneBinder — APPROVED** [review]
